### PR TITLE
Alcohol sleep

### PR DESCRIPTION
**Pull request by kalane15**
## About the PR

Alcohol makes people sleep when they have drunk too much. They can't wake up or be woken up while the effect of drunkenness lasts.
## Why / Balance

About 90u of strong alcohol (I used vodka) lead to sleep human. Sleep duration has min and max values, min value is 15 seconds now and max value is 30 seconds. 1u of strong alcohol prolongs sleep by about 2.2 seconds.

## Technical details

This system uses status system to define how much alcohol entity has drunk. Every alcohol adds current value of time to "drunk" status, so the more alcohol entity has drunk, the longer status effect lasts. Time of effect is used to find out if entity has drunk enough to fall asleep and to calculate how long entity will sleep. 

## Media

- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- tweak: Alcohol can make you sleep if you drink too much. You will sleep until you sober up, so be careful, because big amount of alcohol can disable you for a very long time.

### CODE REVIEW

**Comment by Eden077** [comment]
Finally, a way to handle drunk patrons without moving them by force

**Comment by Partmedia** [comment]
In your PR description, please add discussion about:

- Balance, particularly how many units of alcohol lead to sleep, and how long sleep lasts, and balance implications
- Technical details, i.e. how you implemented this

File: Content.Client/Drunk/DrunkSystem.cs [hunk 1]
```diff
@@ -2,16 +2,17 @@
...
 
 public sealed class DrunkSystem : SharedDrunkSystem
 {
     [Dependency] private readonly IPlayerManager _player = default!;
     [Dependency] private readonly IOverlayManager _overlayMan = default!;
-
+    [ValidatePrototypeId<StatusEffectPrototype>]
+    private const string StatusEffectKey = "ForcedSleep";
     private DrunkOverlay _overlay = default!;
-
```
Comment (ficcialfaint) [inline Content.Client/Drunk/DrunkSystem.cs:14]: and here

File: Content.Server/Drunk/DrunkSystem.cs [hunk 2]
```diff
@@ -1,7 +1,41 @@
...
+            if (StatusEffectsSystem.HasStatusEffect(uid, SleepKey))
+            {
+                if (timeForSleep <= MaxTimeSleep)
+                {
+                    StatusEffectsSystem.TrySetTime(uid, SleepKey, TimeSpan.FromSeconds(timeForSleep));
+                }
+                return;
+            }
+            StatusEffectsSystem.TryAddStatusEffect<ForcedSleepingComponent>(uid, SleepKey, TimeSpan.FromSeconds(MinTimeSleep), false);
+        }
```
Comment (ficcialfaint) [inline Content.Server/Drunk/DrunkSystem.cs:39]: Invert the if-statement

File: Content.Shared/Drunk/DrunkSystem.cs [hunk 3]
```diff
@@ -9,8 +11,33 @@ public abstract class SharedDrunkSystem : EntitySystem
...
     public const string DrunkKey = "Drunk";
 
-    [Dependency] private readonly StatusEffectsSystem _statusEffectsSystem = default!;
+    [ValidatePrototypeId<StatusEffectPrototype>]
+    public const string SleepKey = "ForcedSleep";
+    
+    [Dependency] protected readonly StatusEffectsSystem StatusEffectsSystem = default!;
     [Dependency] private readonly SharedSlurredSystem _slurredSystem = default!;
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
+    public override void Initialize()
```
Comment (ficcialfaint) [inline Content.Shared/Drunk/DrunkSystem.cs:20]:
```suggestion
    [Dependency] private readonly IGameTiming _gameTiming = default!;

    public override void Initialize()
```

File: Content.Shared/Drunk/DrunkSystem.cs [hunk 4]
```diff
@@ -9,8 +11,33 @@ public abstract class SharedDrunkSystem : EntitySystem
     [ValidatePrototypeId<StatusEffectPrototype>]
     public const string DrunkKey = "Drunk";
 
-    [Dependency] private readonly StatusEffectsSystem _statusEffectsSystem = default!;
+    [ValidatePrototypeId<StatusEffectPrototype>]
+    public const string SleepKey = "ForcedSleep";
+    
+    [Dependency] protected readonly StatusEffectsSystem StatusEffectsSystem = default!;
```
Comment (ficcialfaint) [inline Content.Shared/Drunk/DrunkSystem.cs:17]: Why is it protected?

File: Content.Shared/Drunk/DrunkSystem.cs [hunk 5]
```diff
@@ -9,8 +11,33 @@ public abstract class SharedDrunkSystem : EntitySystem
...
+        {
+            StatusEffectsSystem.TryRemoveStatusEffect(uid, SleepKey, component);
+            return;
+        }
+
+        if (args.Key == SleepKey)
+        {
+            StatusEffectsSystem.TryRemoveStatusEffect(uid, DrunkKey, component);
+            return;
+        }
```
Comment (ficcialfaint) [inline Content.Shared/Drunk/DrunkSystem.cs:39]: just combine the if-statements and pass the `args.Key` in the `TryRemoveStatusEffect`'s second argument

File: Content.Shared/Drunk/DrunkSystem.cs [hunk 6]
```diff
@@ -20,29 +47,28 @@ public abstract class SharedDrunkSystem : EntitySystem
 
         if (TryComp<LightweightDrunkComponent>(uid, out var trait))
             boozePower *= trait.BoozeStrengthMultiplier;
-
```
Comment (ficcialfaint) [inline Content.Shared/Drunk/DrunkSystem.cs:23]: why?

**Review by ficcialfaint — CHANGES_REQUESTED** [review]

File: Content.Shared/Drunk/DrunkSystem.cs (hunk 4 above)
Comment (kalane15) [reply Content.Shared/Drunk/DrunkSystem.cs:17]: is is used by server DrunkSystem, which inherits from Shared

File: Content.Shared/Drunk/DrunkSystem.cs (hunk 5 above)
Comment (kalane15) [reply Content.Shared/Drunk/DrunkSystem.cs:40]: but it doesn't remove status with args.Key, it removes the other status

File: Content.Server/Drunk/DrunkSystem.cs (hunk 2 above)
Comment (kalane15) [reply Content.Server/Drunk/DrunkSystem.cs:39]: what statement do you mean?

Comment (ficcialfaint) [reply Content.Server/Drunk/DrunkSystem.cs:39]: On L28

File: Content.Shared/Drunk/DrunkSystem.cs (hunk 5 above)
Comment (ficcialfaint) [reply Content.Shared/Drunk/DrunkSystem.cs:40]:
```suggestion
        if (args.Key == DrunkKey || args.Key == SleepKey)
        {
            StatusEffectsSystem.TryRemoveStatusEffect(uid, args.Key, component);
            return;
        }
```

**Pull request closed without merging**